feature: Active Directory user in nested groups

MrDragon opened the pull request:
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/librenms/librenms?pullRequest=7175) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/librenms/librenms?pullRequest=7175) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/librenms/librenms?pullRequest=7175) it.</sub>

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/3ecafb81-fee1-4ddc-a060-bb9e5b5171e6) was created.

MrDragon commented on html/includes/authentication/active_directory.inc.php:
```diff
@@ -37,33 +37,43 @@ function authenticate($username, $password)
                 }
 
-                throw new AuthenticationException();
+                return false;
```
Sorry, was my lack of understanding. Reverting.

MrDragon commented on html/includes/authentication/active_directory.inc.php:
```diff
@@ -37,33 +37,43 @@ function authenticate($username, $password)
                 if (isset($config['auth_ad_debug']) && $config['auth_ad_debug']) {
                     if ($entries['count'] == 0) {
-                        throw new AuthenticationException('No groups found for user, check base dn');
-                    } else {
```
I'm not searching what group(s) a user directly belongs to, but look at the defined groups and check if the user is a member of the listed groups or subgroups of that group. As far as I know "count" is always 0 or positive. On that search, if it's positive, the user is a member of that group and we've already returned "true", which only leaves "0" as a possibility.
Maybe the message should be more precise, like "User is not in one of the required groups, or the defined groups are outside the base dn" ?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/52a12110-b843-4569-a13d-045165ce8d0b) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/981dd9da-01c0-460d-8d9a-141eb1cd4bd4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/52a12110-b843-4569-a13d-045165ce8d0b) completed: **1 new issues, 2 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/981dd9da-01c0-460d-8d9a-141eb1cd4bd4) completed: **1 new issues, 2 updated code elements**

murrant commented on html/includes/authentication/active_directory.inc.php:
```diff
@@ -104,6 +90,38 @@ function reauthenticate($sess_id, $token)
+        array("DN")
+    );
+    $entries = ldap_get_entries($ldap_connection, $search);
+    unset($entries[0]['memberof']['count']); //remove the annoying count
```
The unset doesn't seem to be needed.

murrant commented on html/includes/authentication/active_directory.inc.php:
```diff
@@ -104,6 +90,38 @@ function reauthenticate($sess_id, $token)
+    global $config, $ldap_connection;
+
+    // get DN for auth_ad_group
+    $filter = "(&(objectClass=group)(cn=$groupname))";
```
$filter is unused

murrant commented:
I know the current code does not use it, but if you could switch to the Config class that would be great (not doing so won't block the PR).

Soo, `$config['auth_ad_groups'][$group]['level']` -> `Config::get("auth_ad_groups.$group.level")`

You can also specify a default value in the second argument to return when the variable is unset.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7b2951f7-17cf-46ae-b24d-2f431cc533c7) was created.

MrDragon commented:
Ok, fixed issues (I hope), and changed that line to Config::get.
I didn't change other "$config" references, because, well, tbh, I don't know what I'm doing. Hadn't used classes before, haven't used 'git' before. I'm just an old-school procedural language, "patch -p0" kinda guy, good enough to mess with code until it does what I need it to do ;)

I appreciate your patience and input!

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7b2951f7-17cf-46ae-b24d-2f431cc533c7) completed: **2 updated code elements**

MrDragon commented:
Sorry if I put in too many pull requests - I think I may have finally figured out where to see if I already requested that or not. Not quite sure though...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/daf67c82-b4ba-453d-91d5-debd5f158b8b) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/daf67c82-b4ba-453d-91d5-debd5f158b8b) completed: **2 updated code elements**

murrant commented:
Nice work @MrDragon :)

laf commented:
@MrDragon We had to revert this PR due to reported issues of AD auth problems: #7254

murrant commented:
Part of the issue is Config::get(), because some people have periods in groups and we use the group name as the key in the config.

Sorry about that.  Just using $config[] will fix that.  I'm not sure if we have other issues though.

MrDragon commented:
Thank you for re-fixing that.
